core/templates/components/graph-services/graph-layout.service.spec.ts

```diff
@@ -0,0 +1,1251 @@
+    let finalNodeIds: string[] = ['End'];
+    let links = [];
+
+    for (let i = 1; i <= MAX_NODES_PER_ROW + 1; i++) {
+      let stateName = 'State' + (i + 1);
+      nodes[stateName] = stateName;
+
+      links.push({
+        source: 'State0',
+        target: stateName
+      });
+      links.push({
+        source: stateName,
+        target: 'End'
+      });
+    }
+
+    let returnedLayoutNodeData = sgls.computeLayout(
+      nodes, links, initNodeId, finalNodeIds);
+    let countNodesDepthOne: number = 0;
+    for (let nodeId in nodes) {
+      if (returnedLayoutNodeData[nodeId].depth === 1) {
+        countNodesDepthOne++;
+      }
+    }
+
+    expect(countNodesDepthOne).toEqual(MAX_NODES_PER_ROW);
+  });
+
+  it('should place orhpaned node at max depth while computing layout', () => {
+    let nodes = {
+      End: 'End',
+      State0: 'State0',
+      Orphan: 'Orphan',
+      State1: 'State1',
+      State2: 'State2',
+      State3: 'State3',
+      State4: 'State4',
+      State5: 'State5'
+    };
+
+    let links = [
+      {
+        source: 'State5',
+        target: 'End'
+      },
+      {
+        source: 'State4',
+        target: 'End'
+      },
+      {
+        source: 'State3',
+        target: 'End'
+      },
+      {
+        source: 'State2',
+        target: 'End'
+      },
+      {
+        source: 'State1',
+        target: 'End'
+      },
+      {
+        source: 'State0',
+        target: 'State1'
+      },
+      {
+        source: 'State0',
+        target: 'State2'
+      },
+      {
+        source: 'State0',
+        target: 'State3'
+      },
+      {
+        source: 'State0',
+        target: 'State4'
+      },
+      {
+        source: 'State0',
+        target: 'State5'
+      },
+      {
+        source: 'State0',
+        target: 'State0'
+      }
+    ];
+    let initNodeId = 'State0';
+    let finalNodeIds = ['End'];
+
+    let returnedLayout = sgls.computeLayout(
+      nodes, links, initNodeId, finalNodeIds);
+
+    expect(returnedLayout.End.depth).toBe(3);
+    expect(returnedLayout.Orphan.depth).toBe(4);
+  });
+
+  it('should get last computed layout', () => {
+    let nodes: GraphNodes = {
+      State1: 'State1',
+      State2: 'State2',
+      State3: 'State3',
+      State4: 'State4',
+      State5: 'State5',
+      State6: 'State6',
+      State7: 'State7',
+      State8: 'State8',
+      State9: 'State9',
+      Orphaned: 'Orphaned'
+    };
+
+    let initNodeId: string = 'State1';
+    let finalNodeIds: string[] = ['State4'];
+
+    expect(sgls.getLastComputedArrangement()).toBe(null);
+
+    let computedLayout = sgls.computeLayout(
+      nodes, links1, initNodeId, finalNodeIds);
+
+    expect(sgls.getLastComputedArrangement()).toEqual(computedLayout);
+  });
+
+  it('should return graph boundaries with width less than equal to' +
+    ' maximum allowed graph width', () => {
+    // Here, nodeDataWithPositionValueInPixel1 ans 2 has position values
```

Comment:
I don't understand this comment ("1 ans 2"?), please rewrite.

If it means "and" then what is the significance of 3 being left out?

You might want to consider including graphical representations so that these graphs can be easily visualized by the reader.

Reply:
Changed the comment, added visualization